src/commands/cmd_query.c

```diff
@@ -67,41 +75,37 @@ void _MGraph_Query(void *args) {
     ResultSet* resultSet = NULL;
     const char *graph_name = RedisModule_StringPtrLen(qctx->graphName, NULL);
 
+    // If this is a write query, acquire write lock.
+    if(AST_ReadOnly(ast)) MGraph_AcquireReadLock();
+    else MGraph_AcquireWriteLock(ctx);
```

Comment:
@swilly22 
There is a strange race condition in this code when we don't lock the context (any read-only query). The context might be freed by one thread, then accessed by another, or the result set can turn out oddly in the concurrent read tests (I assume because replies are being sent to the wrong context.)

I can't figure out why this issue didn't manifest before these changes, though - when can the context change?

Example of an access violation:
```
==3420== Thread 12 thread-pool-7:
==3420== Invalid read of size 1
==3420==    at 0x1A6CA4: moduleGetReplyClient (module.c:1004)
==3420==    by 0x1A6CA4: RM_ReplyWithArray (module.c:1072)
==3420==    by 0x848D91C: _ResultSet_SetupReply (resultset.c:67)
==3420==    by 0x848EB47: NewResultSet (resultset.c:312)
==3420==    by 0x845AC27: NewExecutionPlan (execution_plan.c:240)
==3420==    by 0x849BC99: _MGraph_Query (cmd_query.c:103)
==3420==    by 0x8493E6D: thread_do (thpool.c:337)
==3420==    by 0x53E56DA: start_thread (pthread_create.c:463)
==3420==    by 0x571E88E: clone (clone.S:95)
==3420==  Address 0xcc13e60 is 48 bytes inside a block of size 112 free'd
==3420==    at 0x216B3C: je_valgrind_freelike_block (valgrind.c:33)
==3420==    by 0x849BD79: _MGraph_Query (cmd_query.c:122)
==3420==    by 0x8493E6D: thread_do (thpool.c:337)
==3420==    by 0x53E56DA: start_thread (pthread_create.c:463)
==3420==    by 0x571E88E: clone (clone.S:95)
==3420==  Block was alloc'd at
==3420==    at 0x1DBDC6: je_malloc (jemalloc.c:1441)
==3420==    by 0x143DCD: zmalloc (zmalloc.c:98)
==3420==    by 0x1A72C2: RM_GetThreadSafeContext (module.c:3652)
==3420==    by 0x849BB2D: _MGraph_Query (cmd_query.c:73)
==3420==    by 0x8493E6D: thread_do (thpool.c:337)
==3420==    by 0x53E56DA: start_thread (pthread_create.c:463)
==3420==    by 0x571E88E: clone (clone.S:95)
```

Reply:
Fixed this - it's unsafe to store the `RedisModuleCtx` as a GraphContext member.